README.md
[![exponential_back_off](https://github.com/Nidal-Bakir/exponential_back_off/actions/workflows/exponential_back_off.yml/badge.svg)](https://github.com/Nidal-Bakir/exponential_back_off/actions/workflows/exponential_back_off.yml) [![codecov](https://codecov.io/gh/Nidal-Bakir/exponential_back_off/branch/main/graph/badge.svg?token=WMK1RZQ2JE)](https://codecov.io/gh/Nidal-Bakir/exponential_back_off) [![exponential_back_off on pub.dev](https://img.shields.io/pub/v/exponential_back_off.svg)](https://pub.dev/packages/exponential_back_off)

## Retry failing processes like HTTP requests using an exponential interval between each retry

**Exponential backoff algorithm:**

- An exponential backoff algorithm retries requests exponentially,
 increasing the waiting time between retries up to a maximum backoff time.

<img width="512" src="https://raw.githubusercontent.com/Nidal-Bakir/exponential_back_off/main/digram_image.png"/>

## Features

- [X] Start process
- [X] Stop process
- [X] Reset Process
- [X] Set max attempts
- [X] Set max elapsed time
- [X] Set max delay between retries
- [X] Conditional retry
- [X] On retry callback
- [X] Tweak the exponential delay parameters
- [X] Specify the amount of randomness for the delays
- [X] Custom delay algorithm (inherit from the base class `BackOff`)

## Getting started

 1. Add the package to your `pubspec.yaml`

   ```YAML
   dependencies:
    exponential_back_off: ^x.y.z
   ```

2. Import `exponential_back_off` in a dart file.

```dart
import 'package:exponential_back_off/exponential_back_off.dart';
```

## Usage

- Create `ExponentialBackOff` object

```dart

final exponentialBackOff = ExponentialBackOff();
```

- Make a request

```dart
  final result = await exponentialBackOff.start<Response>(
    () => http.get(Uri.parse('https://www.gnu.org/')),
  );
```

- Handle the result\
    You can handel the result in two ways:
  - By checking if the result `isLeft` or `isRight`. and get the value accordingly.
  - Using the fold function `result.fold((error){},(data){})`. The fold function
     will call the first(Left) function if the result is error otherwise will call second
     function(Right) if the result is data.

   *The error will always be in **Left** and the data will always be in **Right***

  1. **Using if check:**

   ```dart
  result.fold(
    (error) {
      //Left(Exception): handel the error
      print(error);
    },
    (response) {
      //Right(Response): handel the result
      print(response.body);
    },
  );
  ```

  2. **Using fold:**

   ```dart
  result.fold(
    (error) {
      //Left(Exception): handel the error
      print(error);
    },
    (response) {
      //Right(Response): handel the result
      print(response.body);
    },
  );
    ```

#### NOTE

With the default configuration it will be retried up-to 10 times,
sleeping 1st, 2nd, 3rd, ..., 9th attempt: (will not sleep the 10th)

      randomPercent: >=0.0% <=15%
      
      1.  400 ms     +/- (randomPercent of 400 ms)
      2.  800 ms     +/- (randomPercent of 800 ms)
      3.  1600 ms    +/- (randomPercent of 1600 ms)
      4.  3200 ms    +/- (randomPercent of 3200 ms)
      5.  6400 ms    +/- (randomPercent of 6400 ms)
      6.  12800 ms   +/- (randomPercent of 12800 ms)
      7.  25600 ms   +/- (randomPercent of 25600 ms)
      8.  51200 ms   +/- (randomPercent of 51200 ms)
      9.  102400 ms  +/- (randomPercent of 102400 ms)
      10. 204800 ms  +/- (randomPercent of 204800 ms) **will not sleep it**

## Examples

**Because we love to see examples in the README (:**

- Simple use case with the default configuration:

```dart
  final exponentialBackOff = ExponentialBackOff();

  /// The result will be of type [Either<Exception, Response>]
  final result = await exponentialBackOff.start<Response>(
    // Make a request
    () {
      return get(Uri.parse('https://www.gnu.org/'))
          .timeout(Duration(seconds: 10));
    },
    // Retry on SocketException or TimeoutException and other then that the process
    // will stop and return with the error
    retryIf: (e) => e is SocketException || e is TimeoutException,
  );

  /// You can handel the result in two ways
  /// * By checking if the result `isLeft` or `isRight`. and get the value accordingly.
  /// * Using the fold function `result.fold((error){},(data){})`. will call the
  ///   first(Left) function if the result is error otherwise will call second
  ///   function(Right) if the result is data.
  ///
  /// The error will always be in Left and the data will always be in Right

  // using if check
  if (result.isLeft()) {
    //Left(Exception): handel the error
    final error = result.getLeftValue();
    print(error);
  } else {
    //Right(Response): handel the result
    final response = result.getRightValue();
    print(response.body);
  }
  
  // using fold:
  result.fold(
    (error) {
      //Left(Exception): handel the error
      print(error);
    },
    (response) {
      //Right(Response): handel the result
      print(response.body);
    },
  );
```

- Reusing the same object with the default configuration:
  
```dart
  // reset will call stop() and reset everything to zero
  await exponentialBackOff.reset();

  print('interval: ' + exponentialBackOff.interval.toString());
  print('max randomization factor: ' +
      exponentialBackOff.maxRandomizationFactor.toString());
  print('max attempts: ' + exponentialBackOff.maxAttempts.toString());
  print('max delay: ' + exponentialBackOff.maxDelay.toString());
  print('max elapsed time: ' + exponentialBackOff.maxElapsedTime.toString());
  print('==================================================================');

 await exponentialBackOff.start(
    () => get(Uri.parse('https://www.gnu.org/')).timeout(
       Duration.zero, // it will always throw TimeoutException
    ),
    retryIf: (e) => e is SocketException || e is TimeoutException,
    onRetry: (error) {
      print('attempt: ' + exponentialBackOff.attemptCounter.toString());
      print('error: ' + error.toString());
      print('current delay: ' + exponentialBackOff.currentDelay.toString());
      print('elapsed time: ' + exponentialBackOff.elapsedTime.toString());
      print('--------------------------------------------------------');
    },
  );
```

- Tweaks the exponential delay parameters:

```dart
  // Will be retried up-to 5 times,
  // sleeping 1st, 2nd, 3rd, ..., 4th attempt: (will not sleep the 5th)
  //
  //   randomPercent: 0.0%
  //
  //  1. 200 ms    
  //  2. 400 ms     
  //  3. 800 ms    
  //  4. 1600 ms    
  //  5. 3200 ms   **will not sleep it**
 
  final exponentialBackOff = ExponentialBackOff(
    interval: Duration(milliseconds: 100),
    maxAttempts: 5,
    maxRandomizationFactor: 0.0,
    maxDelay: Duration(seconds: 15),
  );

  print('interval: ' + exponentialBackOff.interval.toString());
  print('max randomization factor: ' +
      exponentialBackOff.maxRandomizationFactor.toString());
  print('max attempts: ' + exponentialBackOff.maxAttempts.toString());
  print('max delay: ' + exponentialBackOff.maxDelay.toString());
  print('max elapsed time: ' + exponentialBackOff.maxElapsedTime.toString());
  print('==================================================================');

  await exponentialBackOff.start(
    () => get(Uri.parse('https://www.gnu.org/')).timeout(
      Duration.zero, // it will always throw TimeoutException
    ),
    retryIf: (e) => e is SocketException || e is TimeoutException,
    onRetry: (error) {
      print('attempt: ' + exponentialBackOff.attemptCounter.toString());
      print('error: ' + error.toString());
      print('current delay: ' + exponentialBackOff.currentDelay.toString());
      print('elapsed time: ' + exponentialBackOff.elapsedTime.toString());
      print('--------------------------------------------------------');
    },
  );

```

- Custom Delay
  - create a subclass from Backoff base class

      ```dart
      /// linier delays:
      ///
      ///  1. 1000 ms
      ///  2. 2000 ms
      ///  3. 3000 ms
      ///  4. 4000 ms
      ///  5. 5000 ms
      class CustomDelay extends BackOff {
        CustomDelay({
          super.maxAttempts,
          super.maxDelay,
          super.maxElapsedTime,
        }) : assert(maxAttempts != null || maxElapsedTime != null,
                  'Can not have both maxAttempts and maxElapsedTime null');

        @override
        Duration computeDelay(int attempt, Duration elapsedTime) {
          return Duration(seconds: attempt);
        }
      }
      
      ```

  - Use it as you normally do

      ```dart
        // Will be retried up-to 5 times,
        // sleeping 1st, 2nd, 3rd, ..., 4th attempt: (will not sleep the 5th)
        //
        //  1. 1000 ms
        //  2. 2000 ms
        //  3. 3000 ms
        //  4. 4000 ms
        //  5. 5000 ms   **will not sleep it**

        final customDelay = CustomDelay(maxAttempts: 5);

        print('max attempts: ' + customDelay.maxAttempts.toString());
        print('max delay: ' + customDelay.maxDelay.toString());
        print('max elapsed time: ' + customDelay.maxElapsedTime.toString());
        print('==================================================================');

        await customDelay.start(
          () => get(Uri.parse('https://www.gnu.org/')).timeout(
            Duration.zero, // it will always throw TimeoutException
          ),
          retryIf: (e) => e is SocketException || e is TimeoutException,
          onRetry: (error) {
            print('attempt: ' + customDelay.attemptCounter.toString());
            print('error: ' + error.toString());
            print('current delay: ' + customDelay.currentDelay.toString());
            print('elapsed time: ' + customDelay.elapsedTime.toString());
            print('--------------------------------------------------------');
          },
        );
      ```


CHANGELOG.md
## v0.1.0+2

- Improve the docs

## v0.1.0+1

- Improve the docs

## v0.1.0

- `elapsedTime` now will return real-time duration rather than folding delays
- Improve the docs

## v0.0.3

- Improve the docs

## v0.0.2

- Add tests
- Fix: calling reset() and reusing the backoff object with start() result in inconsistent behavior
- Add examples
- Edit the README

## v0.0.1

- Initial release.


test/either_test.dart
import 'package:test/test.dart';
import 'package:exponential_back_off/src/either.dart';

void main() {
  test('test Left Either', () {
    // arrange
    final error = Exception('error');
    Either<Exception, int> leftEither = Left(error);

    // act
    final isLeft = leftEither.isLeft();
    final isRight = leftEither.isRight();

    final getLeftLeftEither = leftEither.getLeft;
    final getLeftRightEither = leftEither.getRight;

    final getLeftValue = leftEither.getLeftValue;
    final getRightValue = leftEither.getRightValue;

    final foldResult = leftEither.fold(
      (l) => l,
      (r) => 'NO OP',
    );

    // assert
    expect(isLeft, isTrue);
    expect(isRight, isFalse);

    expect(getLeftLeftEither(), equals(leftEither));
    expect(() => getLeftRightEither(), throwsA(isA<NotRightException>()));

    expect(getLeftValue(), equals(error));
    expect(() => getRightValue(), throwsA(isA<NotRightException>()));

    expect(foldResult, equals(error));
  });

  test('test Right Either', () {
    // arrange
    final data = 1;
    Either<Exception, int> rightEither = Right(data);

    // act
    final isLeft = rightEither.isLeft();
    final isRight = rightEither.isRight();

    final getLeftLeftEither = rightEither.getLeft;
    final getLeftRightEither = rightEither.getRight;

    final getLeftValue = rightEither.getLeftValue;
    final getRightValue = rightEither.getRightValue;

    final foldResult = rightEither.fold(
      (l) => 'NO OP',
      (r) => r,
    );

    // assert
    expect(isLeft, isFalse);
    expect(isRight, isTrue);

    expect(() => getLeftLeftEither(), throwsA(isA<NotLeftException>()));
    expect(getLeftRightEither(), equals(rightEither));

    expect(getRightValue(), equals(data));
    expect(() => getLeftValue(), throwsA(isA<NotLeftException>()));

    expect(foldResult, equals(data));
  });
}


test/exponential_back_off_test.dart
import 'dart:math';

import 'package:test/test.dart';

import 'package:exponential_back_off/exponential_back_off.dart';

class TestException implements Exception {
  final dynamic message;

  TestException([this.message]);

  @override
  String toString() {
    Object? message = this.message;
    if (message == null) return "TextException";
    return "TextException: $message";
  }
}

void main() {
  test('attemptCounter should equals maxAttempts', () async {
    // arrange
    final maxAttempts = 3;
    final expo = ExponentialBackOff(
      maxAttempts: maxAttempts,
      interval: Duration(milliseconds: 200),
      maxDelay: Duration(seconds: 10),
      maxRandomizationFactor: 0.15,
    );
    int callCounter = 0;

    // act
    await expo.start(() {
      ++callCounter;
      throw TestException('test throw');
    });

    // assert
    expect(expo.attemptCounter, equals(maxAttempts));
    expect(callCounter, equals(maxAttempts));
  });

  test('elapsedTime should be less then or equal to maxElapsedTime', () async {
    // arrange
    final maxElapsedTime = Duration(seconds: 3);
    final expo = ExponentialBackOff(
      maxElapsedTime: maxElapsedTime,
      interval: Duration(milliseconds: 200),
      maxDelay: Duration(seconds: 10),
      maxRandomizationFactor: 0.15,
    );

    // act
    await expo.start(() => throw TestException('test throw'));

    // assert
    expect(expo.elapsedTime, lessThanOrEqualTo(maxElapsedTime));
  });

  test(
      'elapsedTime should be equal to expected precise sleep time '
      'when maxRandomizationFactor equal 0 ', () async {
    // arrange
    final interval = Duration(milliseconds: 50);
    final maxAttempts = 3;
    final expo = ExponentialBackOff(
      maxAttempts: maxAttempts,
      interval: interval,
      maxDelay: Duration(seconds: 10),
      maxRandomizationFactor: 0,
    );

    int callCounter = 0;

    // act
    await expo.start(() {
      ++callCounter;
      throw TestException('test throw');
    });

    // assert
    final expectedPreciseSleepTimeMillis =
        (interval * pow(2, 1) + interval * pow(2, 2)).inMilliseconds;

    expect(
      expo.elapsedTime.inMilliseconds,
      inInclusiveRange(
        expectedPreciseSleepTimeMillis,
        expectedPreciseSleepTimeMillis +
            100, // + ~100 the natural processing time of the code
      ),
    );
    expect(callCounter, equals(maxAttempts));
    expect(expo.isProcessRunning(), isFalse);
  });

  test(
      'the computed delay should be equal to maxDelay in case the computed '
      'delay is grater than the specified maxDelay', () {
    // arrange
    final interval = Duration(seconds: 99);
    final maxDelay = Duration(seconds: 5);
    final expo = ExponentialBackOff(
      maxAttempts: 3,
      interval: interval,
      maxDelay: maxDelay,
      maxRandomizationFactor: 0.15,
    );

    // act
    final delay = expo.computeDelay(99, Duration.zero);

    // assert
    expect(delay, equals(maxDelay));
  });

  test(
      'computeDelay() should output delays in InclusiveRange[delay-rand,delay+rand] '
      'which follow the defined formula: delay = interval * 2^min(attempt,31) '
      '+/- (random percent`>= 0.0 and <=[randomizationFactor]`of that delay',
      () {
    final interval = const Duration(milliseconds: 200);
    final maxDelay = const Duration(seconds: 30);
    final maxRandFactor = 0.15;
    final expo = ExponentialBackOff(
      interval: interval,
      maxDelay: maxDelay,
      maxRandomizationFactor: maxRandFactor,
    );

    for (var i = 1; i <= 1000; i++) {
      // act
      final delay = expo.computeDelay(i, Duration.zero);

      // assert
      final exp = min(31, i);
      final expectedRawDelay = interval * pow(2, exp);
      var maxRandomDelay = expectedRawDelay + expectedRawDelay * maxRandFactor;
      var minRandomDelay = expectedRawDelay - expectedRawDelay * maxRandFactor;
      maxRandomDelay = maxRandomDelay > maxDelay ? maxDelay : maxRandomDelay;
      minRandomDelay = minRandomDelay > maxDelay ? maxDelay : minRandomDelay;

      expect(
        delay.inMilliseconds,
        inInclusiveRange(
          minRandomDelay.inMilliseconds,
          maxRandomDelay.inMilliseconds,
        ),
      );
    }
  });

  test(
      'calling stop() function should produce consistent result i.e: '
      'the value of CurrentDelay before calling stop() should be the same after '
      'calling stop() and the same applied for attemptCounter', () async {
    // arrange
    final expo = ExponentialBackOff(
      maxAttempts: 10,
      interval: const Duration(milliseconds: 200),
      maxDelay: const Duration(seconds: 30),
      maxRandomizationFactor: 0.15,
    );
    int callCounter = 0;

    // act
    expo.start(() {
      ++callCounter;
      throw TestException('test throw');
    });
    await Future.delayed(Duration(seconds: 2));

    final beforeStopCurrentDelay = expo.currentDelay;
    final beforeStopAttemptCounter = expo.attemptCounter;
    await expo.stop();
    final afterStopCurrentDelay = expo.currentDelay;
    final afterStopAttemptCounter = expo.attemptCounter;

    // assert
    expect(
      expo.isProcessStopped(),
      isTrue,
      reason: 'The process should be stopped',
    );

    expect(
      afterStopCurrentDelay,
      equals(beforeStopCurrentDelay),
      reason: 'CurrentDelay should be the same before and after the stop()',
    );
    expect(
      afterStopAttemptCounter,
      equals(beforeStopAttemptCounter),
      reason: 'AttemptCounter should be the same before and after the stop()',
    );

    expect(
      callCounter,
      equals(expo.attemptCounter),
    );
  });

  test(
      'calling reset() should call stop() function to stop the current process '
      'and then reset all the vars to its init value (zero)', () async {
    // arrange
    final expo = ExponentialBackOff(
      maxAttempts: 10,
      interval: const Duration(milliseconds: 200),
      maxDelay: const Duration(seconds: 30),
      maxRandomizationFactor: 0.15,
    );

    // act
    expo.start(() {
      throw TestException('test throw');
    });
    await Future.delayed(Duration(seconds: 2));

    await expo.reset();

    // assert
    expect(
      expo.isProcessStopped(),
      isTrue,
      reason: 'The process should be stopped',
    );

    expect(
      expo.elapsedTime,
      equals(Duration.zero),
      reason: 'The elapsedTime should be reset to Duration.zero',
    );
    expect(
      expo.attemptCounter,
      equals(0),
      reason: 'The attemptCounter should be reset to zero',
    );
  });

  test(
      'should return [Left] with [TextException] object as a value when calling '
      'start() and throwing [TextException] in its body', () async {
    // arrange
    final expo = ExponentialBackOff(
      maxAttempts: 3,
      interval: const Duration(milliseconds: 200),
      maxDelay: const Duration(seconds: 30),
      maxRandomizationFactor: 0.15,
    );

    final exception = TestException('test throw');

    // act
    final result = await expo.start(() {
      throw exception;
    });

    // assert
    expect(result, isA<Left>());
    expect(result.getLeftValue(), equals(exception));
    expect(expo.attemptCounter, equals(3));
  });

  test(
      'should return [Left] with [IgnoredProcess] object as a value when calling '
      'stop() while proc() function still working and not in back off delay',
      () async {
    // arrange
    final expo = ExponentialBackOff(
      maxAttempts: 3,
      interval: const Duration(milliseconds: 200),
      maxDelay: const Duration(seconds: 30),
      maxRandomizationFactor: 0.15,
    );

    // act
    final result = expo.start(() async {
      await Future.delayed(Duration(seconds: 1));
    });
    await Future.delayed(Duration(milliseconds: 200));
    await expo.stop();

    // assert
    final res = (await result);
    expect(res, isA<Left>());
    expect(res.getLeftValue(), isA<IgnoredProcess>());
    expect(expo.attemptCounter, equals(1));
  });

  test(
      'should return [Right] with [int] as a value when calling '
      'start() and returning [int] in its body', () async {
    // arrange
    final expo = ExponentialBackOff(
      maxAttempts: 10,
      interval: const Duration(milliseconds: 200),
      maxDelay: const Duration(seconds: 30),
      maxRandomizationFactor: 0.15,
    );

    final returnValue = 10;

    // act
    final result = await expo.start<int>(() async {
      await Future.delayed(Duration(milliseconds: 500));
      return returnValue;
    });

    // assert
    expect(result, isA<Right>());
    expect(result.getRightValue(), equals(returnValue));
    expect(expo.attemptCounter, equals(1));
  });

  test(
      'should return [Right] with [int] as a value when calling '
      'start() and returning [int] in its body with falling in the first time',
      () async {
    // arrange
    final expo = ExponentialBackOff(
      maxAttempts: 10,
      interval: const Duration(milliseconds: 200),
      maxDelay: const Duration(seconds: 30),
      maxRandomizationFactor: 0.15,
    );

    final returnValue = 10;
    int callCounter = 0;
    final exception = TestException('test throw');

    // act
    final result = await expo.start<int>(() async {
      ++callCounter;

      await Future.delayed(Duration(milliseconds: 300));

      if (callCounter == 2) {
        return returnValue;
      }

      throw exception;
    });

    // assert
    expect(result, isA<Right>());
    expect(result.getRightValue(), equals(returnValue));

    expect(expo.attemptCounter, equals(2));
    expect(callCounter, equals(2));
    expect(expo.isProcessStopped(), isTrue);
  });
  test('verify that onRetry() function get called on every attempt to retry',
      () async {
    // arrange
    final maxAttempts = 3;
    final expo = ExponentialBackOff(
      maxAttempts: maxAttempts,
      interval: Duration(milliseconds: 200),
      maxDelay: Duration(seconds: 10),
      maxRandomizationFactor: 0.15,
    );

    int callCounter = 0;
    int onRetryCallCounter = 0;

    // act
    await expo.start(
      () {
        ++callCounter;
        throw TestException('test throw');
      },
      onRetry: (error) {
        ++onRetryCallCounter;
      },
    );

    // assert
    expect(expo.attemptCounter, equals(maxAttempts));
    expect(callCounter, equals(expo.attemptCounter));
    expect(onRetryCallCounter, equals(expo.attemptCounter));
  });
  test(
      'verify that onRetry() and retryIf() functions not called when the process '
      'success from the first time', () async {
    // arrange
    final maxAttempts = 3;
    final expo = ExponentialBackOff(
      maxAttempts: maxAttempts,
      interval: Duration(milliseconds: 200),
      maxDelay: Duration(seconds: 10),
      maxRandomizationFactor: 0.15,
    );

    int callCounter = 0;
    int onRetryCallCounter = 0;
    int retryIfCallCounter = 0;
    int returnValue = 10;

    // act
    await expo.start(
      () async {
        ++callCounter;
        return returnValue;
      },
      retryIf: (error) {
        ++retryIfCallCounter;
        return true;
      },
      onRetry: (error) {
        ++onRetryCallCounter;
      },
    );

    // assert
    expect(expo.attemptCounter, equals(1));
    expect(callCounter, equals(expo.attemptCounter));
    expect(onRetryCallCounter, equals(0));
    expect(retryIfCallCounter, equals(0));
  });

  test(
      'verify that retryIf() break the loop and stop the process if it returned false',
      () async {
    // arrange
    final maxAttempts = 3;
    final expo = ExponentialBackOff(
      maxAttempts: maxAttempts,
      interval: Duration(milliseconds: 200),
      maxDelay: Duration(seconds: 10),
      maxRandomizationFactor: 0.15,
    );

    int callCounter = 0;
    int retryIfCallCounter = 0;

    final exception = TestException('test throw');
    // act
    final result = await expo.start(
      () async {
        ++callCounter;
        throw exception;
      },
      retryIf: (error) {
        ++retryIfCallCounter;
        return false;
      },
    );

    // assert
    expect(expo.attemptCounter, equals(1));
    expect(callCounter, equals(1));
    expect(retryIfCallCounter, equals(1));

    expect(result, isA<Left>());
    expect(result.getLeftValue(), isA<TestException>());
    expect(result.getLeftValue(), equals(exception));
  });

  test(
      'verify that retryIf() will not break the loop and the process will'
      'continue retrying if it returned true', () async {
    // arrange
    final maxAttempts = 3;
    final expo = ExponentialBackOff(
      maxAttempts: maxAttempts,
      interval: Duration(milliseconds: 200),
      maxDelay: Duration(seconds: 10),
      maxRandomizationFactor: 0.15,
    );

    int callCounter = 0;
    int retryIfCallCounter = 0;

    final exception = TestException('test throw');
    // act
    final result = await expo.start(
      () async {
        ++callCounter;
        throw exception;
      },
      retryIf: (error) {
        ++retryIfCallCounter;
        return true;
      },
    );

    // assert
    expect(expo.attemptCounter, equals(maxAttempts));
    expect(callCounter, equals(expo.attemptCounter));
    // -1 because retryIf() will not be called in the last retry
    expect(retryIfCallCounter, equals(maxAttempts - 1));

    expect(result, isA<Left>());
    expect(result.getLeftValue(), isA<TestException>());
    expect(result.getLeftValue(), equals(exception));
  });

  test(
    'calling reset() and using the object agin should have a consistent behavior, '
    'as it was called for the first time',
    () async {
      // arrange
      final maxAttempts = 3;
      final interval = Duration(milliseconds: 50);
      final maxRandFactor = 0.15;
      final expo = ExponentialBackOff(
        maxAttempts: maxAttempts,
        interval: interval,
        maxDelay: Duration(seconds: 10),
        maxRandomizationFactor: maxRandFactor,
      );

      int callCounter = 0;
      int retryIfCallCounter = 0;

      final exception = TestException('test throw');

      // act
      await expo.start(
        () async {
          ++callCounter;
          throw exception;
        },
        retryIf: (error) {
          ++retryIfCallCounter;
          return true;
        },
      );

      await expo.reset();
      callCounter = 0;
      retryIfCallCounter = 0;

      await expo.start(
        () async {
          ++callCounter;
          throw exception;
        },
        retryIf: (error) {
          ++retryIfCallCounter;
          return true;
        },
      );

      // assert

      expect(
        expo.isProcessRunning(),
        isFalse,
        reason: 'The process should be stopped',
      );

      var expectedRawDelay = interval * pow(2, 1);
      var maxRandomDelay = expectedRawDelay + expectedRawDelay * maxRandFactor;
      var minRandomDelay = expectedRawDelay - expectedRawDelay * maxRandFactor;
      expectedRawDelay = interval * pow(2, 2);
      maxRandomDelay += expectedRawDelay + expectedRawDelay * maxRandFactor;
      minRandomDelay += expectedRawDelay - expectedRawDelay * maxRandFactor;

      expect(
        expo.elapsedTime.inMilliseconds,
        inInclusiveRange(
          minRandomDelay.inMilliseconds,
          maxRandomDelay.inMilliseconds,
        ),
      );

      expect(
        expo.attemptCounter,
        equals(maxAttempts),
        reason: 'The attemptCounter should be 3',
      );

      expect(
        callCounter,
        equals(maxAttempts),
        reason: 'The callCounter should be 3',
      );
      expect(
        retryIfCallCounter,
        equals(maxAttempts - 1),
        reason: 'The retryIfCallCounter should be 3',
      );
    },
  );
}


example/lib/example2.dart
import 'dart:async';
import 'dart:io';

import 'package:exponential_back_off/exponential_back_off.dart';
import 'package:http/http.dart';

void main(List<String> arguments) async {
  // Will be retried up-to 5 times,
  // sleeping 1st, 2nd, 3rd, ..., 4th attempt: (will not sleep the 5th)
  //
  //   randomPercent: 0.0%
  //
  //  1. 200 ms
  //  2. 400 ms
  //  3. 800 ms
  //  4. 1600 ms
  //  5. 3200 ms   **will not sleep it**

  final exponentialBackOff = ExponentialBackOff(
    interval: Duration(milliseconds: 100),
    maxAttempts: 5,
    maxRandomizationFactor: 0.0,
    maxDelay: Duration(seconds: 15),
  );

  print('interval: ' + exponentialBackOff.interval.toString());
  print('max randomization factor: ' +
      exponentialBackOff.maxRandomizationFactor.toString());
  print('max attempts: ' + exponentialBackOff.maxAttempts.toString());
  print('max delay: ' + exponentialBackOff.maxDelay.toString());
  print('max elapsed time: ' + exponentialBackOff.maxElapsedTime.toString());
  print('==================================================================');

  await exponentialBackOff.start(
    () => get(Uri.parse('https://www.gnu.org/')).timeout(
      Duration.zero, // it will always throw TimeoutException
    ),
    retryIf: (e) => e is SocketException || e is TimeoutException,
    onRetry: (error) {
      print('attempt: ' + exponentialBackOff.attemptCounter.toString());
      print('error: ' + error.toString());
      print('current delay: ' + exponentialBackOff.currentDelay.toString());
      print('elapsed time: ' + exponentialBackOff.elapsedTime.toString());
      print('--------------------------------------------------------');
    },
  );
}


lib/src/exponential_back_off.dart
import 'dart:math' as math;

import 'back_off_base.dart';

/// Implementation of [BackOff] that increases the back off period for each
/// retry attempt using a function [computeDelay] that grows exponentially with
/// added/subtracted random value.
///
/// With the default configuration it will be retried up-to 10 times,
/// sleeping 1st, 2nd, 3rd, ..., 9th attempt: (will not sleep the 10th)
///
///   randomPercent: >=0.0% <=15%
///
///  1. 400 ms     +/- (randomPercent of 400 ms)
///  2. 800 ms     +/- (randomPercent of 800 ms)
///  3. 1600 ms    +/- (randomPercent of 1600 ms)
///  4. 3200 ms    +/- (randomPercent of 3200 ms)
///  5. 6400 ms    +/- (randomPercent of 6400 ms)
///  6. 12800 ms   +/- (randomPercent of 12800 ms)
///  7. 25600 ms   +/- (randomPercent of 25600 ms)
///  8. 51200 ms   +/- (randomPercent of 51200 ms)
///  9. 102400 ms  +/- (randomPercent of 102400 ms)
///  10. 204800 ms +/- (randomPercent of 204800 ms) **will not sleep it**
///
/// **Example**
/// ```dart
///   final exponentialBackOff = ExponentialBackOff();
///   /// The result will be of type [Either<Exception, Response>]
///   final result = await exponentialBackOff.start<Response>(
///     // Make a request
///    () {
///     return get(Uri.parse('https://www.gnu.org/'))
///       .timeout(Duration(seconds: 10));
///    },
///    // Retry on SocketException or TimeoutException and other then that the process
///    // will stop and return with the error
///    retryIf: (e) => e is SocketException || e is TimeoutException,
///   );
///
///  result.fold(
///    (error) {
///      //Left(Exception): handel the error
///      print(error);
///    },
///   (response) {
///     //Right(Response): handel the result
///     print(response.body);
///    },
///  );
/// ```
class ExponentialBackOff extends BackOff {
  final _rand = math.Random();

  /// Delay interval that will growth exponentially.
  ///
  /// Defaults to 200 ms, which results in the following delays:
  ///
  ///  1. 400 ms
  ///  2. 800 ms
  ///  3. 1600 ms
  ///  4. 3200 ms
  ///  5. 6400 ms
  ///  6. 12800 ms
  ///  7. 25600 ms
  ///  8. 51200 ms
  ///  9. 102400 ms
  ///  10. 204800 ms
  ///
  /// Before applying [maxRandomizationFactor].
  final Duration interval;

  /// The maximum random percentage of the computed delay to be added/subtracted
  /// to/from the resulted delay, given as fraction between 0 and 1.
  ///
  /// e.g: If `0.15` (default) this will generate a
  /// value percentage between `0.0% and 0.15%` of that delay will be
  /// added/subtracted to/from the resulted delay.
  final double maxRandomizationFactor;

  /// Create exponential delay between retries.
  ///
  /// [maxAttempts]: Maximum retry attempts, defaults to 10 attempts.
  ///
  /// [maxElapsedTime]: Maximum elapsed time while retrying, defaults to null.
  ///
  /// [maxDelay]: Maximum delay between retries, defaults to 30 seconds.
  ExponentialBackOff({
    this.maxRandomizationFactor = 0.15,
    this.interval = const Duration(milliseconds: 200),
    super.maxAttempts = 10,
    super.maxElapsedTime,
    super.maxDelay = const Duration(seconds: 30),
  });

  /// Calculates the next back off interval using the formula:
  ///
  /// delay = interval * 2^min(attempt,31)
  /// +/- (random percent`>= 0.0 and <=[randomizationFactor]`of that delay.
  @override
  Duration computeDelay(int attempt, _) {
    final exp = math.min(attempt, 31); // prevent overflows.

    var delay = interval * math.pow(2, exp);

    // Random value in the range [-randomizationFactor, +randomizationFactor].
    final randomPercent = _rand.nextDouble() * maxRandomizationFactor * 2 -
        maxRandomizationFactor;

    final randomDelay = delay.inMilliseconds * randomPercent;

    delay = Duration(milliseconds: delay.inMilliseconds + randomDelay.ceil());

    return delay > maxDelay ? maxDelay : delay;
  }
}
